program-entrypoint/src/lib.rs

```diff
@@ -281,18 +281,61 @@ macro_rules! custom_panic_default {
 
 /// The bump allocator used as the default rust heap when running programs.
 pub struct BumpAllocator {
+    #[deprecated(
+        since = "2.2.2",
+        note = "This field should not be accessed directly. It will become private in future versions"
+    )]
     pub start: usize,
+    #[deprecated(
+        since = "2.2.2",
+        note = "This field should not be accessed directly. It will become private in future versions"
+    )]
     pub len: usize,
 }
+
+impl BumpAllocator {
+    /// Creates the allocator tied to a provided slice.
+    /// This will not initialize the provided memory, except for the first
+    /// bytes where the pointer is stored.
+    ///
+    /// # Safety
+    /// As long as BumpAllocator or any of its allocations are alive,
+    /// writing into or deallocating the arena will cause UB.
+    ///
+    /// Integer arithmetic in this global allocator implementation is safe when
+    /// operating on the prescribed `HEAP_START_ADDRESS` and `HEAP_LENGTH`. Any
+    /// other use may overflow and is thus unsupported and at one's own risk.
+    #[inline]
+    #[allow(clippy::arithmetic_side_effects)]
+    pub unsafe fn new(arena: &mut [u8]) -> Self {
+        debug_assert!(
+            arena.len() > size_of::<usize>(),
+            "Arena should be larger than usize"
+        );
+
+        // create a pointer to the start of the arena
+        // that will hold an address of the byte following free space
+        let pos_ptr = arena.as_mut_ptr() as *mut usize;
+        // initialize the data there
+        *pos_ptr = pos_ptr as usize + arena.len();
```

Comment:
The change looks good, but for my own understanding, how does this avoid undefined behavior exactly?

Is the idea that the `new` call forces the optimizer to keep `heap` around in the tests because some part of it is written to? Would there still be UB if this code here were removed?

Reply:
fn new() correctly initializes the arena content where the allocated amount is, which was not done before, as the code assumed all arenas come zeroed. If you were to instantiate the BumpAllocator over a non-zeroed slice, it would cause UB the moment you call alloc().

The reason bug UB was triggered before this change was due to `heap` variable not marked as mut, so compiler assumed that it does not need to reinitialize it (since noone would write into it).

Reply:
Ah very interesting, thanks for the explanation!